File: problems/balancedparentheses.cpp

#include <iostream>

using namespace std ;

#define ms 100
class stack
{
	private :
		int arr[ms] ;
		int top = -1 ;
	public :
		bool empty()
		{
			if(top == -1)
				return true ;
			else
				return false ;
		}
		bool full()
		{
			if(top == ms - 1)
				return true ;
			else
				return false ;
		}
		void push(int v)
		{
			if(full())
				cout << "full stack" << endl ;
			else
				top += 1 ;
			arr[top] = v ;
		}
		void pop()
		{
			if(empty())
				cout << "empty stack" << endl ;
			else
				top -= 1 ;
		}
		int peek()
		{
			if(empty())
				return -1 ;
			else
				return arr[top] ;
		}
} ;

stack s ;

bool parentheses(char a , char b)
{
	if((a == '(' and b == ')') or (a == '[' and b == ']') or (a == '{' and b == '}'))
		return true ;
	else
		return false ;
}

bool operation(string e)
{
	for(int i = 0 ; i < e.length() ; i++)
	{
		if(e[i] == ' ')
			continue ;
		else if(e[i] == '(' or e[i] == '[' or e[i] == '{')
			s.push(e[i]) ;
		else
		{
			if (s.empty())
				return false ;
			else if((parentheses(s.peek() , e[i])) == false)
				return false ;
			else
				s.pop() ;
		}
	}

	if (s.empty())
		return true ;
	else
		return false ;
}

int main(void)

{
	string expression ;
	cout << "Enter parentheses : " ;
	getline(cin , expression) ;

	if(operation(expression))
		cout << "Balanced" << endl ;
	else
		cout << "Not Balanced" << endl ;

	return 0 ;
}

File: problems/balancedparentheses.py

class stack() :

	def __init__ (self) :

		self.array = []

	def push(self , value) :

		self.array.append(value)
 
	def pop(self) :

		if (self.isempty()) : return
		self.array.pop()

	def peek(self) :

		return self.array[-1]

	def isempty(self) :

		return len(self.array) == 0

	def show(self) :

		if self.isempty() : print("Empty stack")
		else :

			while (not self.isempty()) :

				print(self.peek() , end = " ")
				self.pop()

		print(" ")


def main() :

	def pair(open , close) :

		if open == '(' and close == ')' : return True
		elif open == '[' and close == ']' : return True
		elif open == '{' and close == '}' : return True
		else : return False

	def balanced(string) :

		_object = stack()

		for i in string :

			if i == '(' or i == '[' or i == '{' : _object.push(i)
			elif i == ')' or i == ']' or i == '}' :

				if (_object.isempty()) : return False
				elif not pair(_object.peek() , i) : return False
				else : _object.pop()

		if (_object.isempty()) : return True
		else : return False

	s = str(input("Enter Parenthese : "))

	return "Balanced" if balanced(s) else "Not Balanced"

if __name__ == '__main__': 

	print(main())
